go/store/blobstore/gcs.go

```diff
@@ -155,3 +162,61 @@ func (bs *GCSBlobstore) CheckAndPut(ctx context.Context, expectedVersion, key st
 
 	return ver, err
 }
+
+func (bs *GCSBlobstore) Contatenate(ctx context.Context, key string, sources []string) (string, error) {
+	// GCS compose has a batch size limit,
+	// recursively compose sources
+	for len(sources) > composeBatch {
+		// compose subsets of |sources| in batches,
+		// store tmp composite objects in |next|
+		var next []string
+		for len(sources) > 0 {
+			tmp := uuid.New().String()
+			batch := min(composeBatch, len(sources))
+			_, err := bs.composeObjects(ctx, tmp, sources[:batch])
+			if err != nil {
+				return "", err
+			}
+			next = append(next, tmp)
+			sources = sources[batch:]
+		}
+		sources = next
+	}
+	return bs.composeObjects(ctx, key, sources)
+}
+
+func (bs *GCSBlobstore) composeObjects(ctx context.Context, composite string, sources []string) (gen string, err error) {
+	if len(sources) > composeBatch {
+		return "", fmt.Errorf("too many objects to compose (%d > %d)", len(sources), composeBatch)
+	}
+
+	var a *storage.ObjectAttrs
+	objects := make([]*storage.ObjectHandle, len(sources))
+	for i := range objects {
+		oh := bs.bucket.Object(path.Join(bs.prefix, sources[i]))
```

Comment:
Again a source for concurrency...maybe less important than doing the top-level composes in parallel.